File: getSpectrum.py
import numpy as np
from scipy.fft import fft, fftfreq, fftshift

def compute_spectrum(signal, sampling_rate=1.0):
    """
    Параметры:
    signal :  Входной одномерный сигнал.
    sampling_rate :  Частота дискретизации сигнала (в Гц). По умолчанию 1.0.

    Возвращает:
    frequencies : ndarray
        Массив частот (в Гц).
    amplitudes : ndarray
        Массив амплитуд соответствующих частот.
    """    
    n = len(signal)
    fft_result = fft(signal)
    amplitudes = np.abs(fft_result) / n
    frequencies = fftshift(fftfreq(n, d=1/sampling_rate))
    amplitudes = fftshift(amplitudes)
    return frequencies, amplitudes

File: MathLab/getSpectrum.m
function [frequencies, amplitudes] = compute_spectrum(signal, sampling_rate)
% compute_spectrum - вычисляет спектр сигнала с центровкой нулевой частоты
%
% Входные параметры:
%   signal        - входной одномерный сигнал (вектор)
%   sampling_rate - частота дискретизации (Гц), по умолчанию 1.0
%
% Выходные параметры:
%   frequencies - вектор частот (Гц)
%   amplitudes  - амплитуды соответствующих частот

    if nargin < 2
        sampling_rate = 1.0;
    end

    n = length(signal);
    fft_result = fft(signal);
    amplitudes = abs(fft_result) / n;

    % Частоты с учетом дискретизации
    frequencies = (-floor(n/2) : ceil(n/2)-1) * (sampling_rate / n);

    % Сдвиг спектра, чтобы нулевая частота была в центре
    amplitudes = fftshift(amplitudes);
end
 
